From 0ba3d60d0fe558df4142487120936179da033394 Mon Sep 17 00:00:00 2001
From: Pan Junzhong <junzhong.pan@spacemit.com>
Date: Fri, 16 Aug 2024 20:14:23 +0800
Subject: [PATCH 1277/1448] k1: f_mass_storage: replace kmalloc with
 alloc_pages with DMA32

Running f_mass_storage on dwc3 cause swiotlb full. alloc_pages
support DMA32 flag, and the FSG_BUFLEN is 16K, applicable for
alloc_pages.

Change-Id: Ic2f33783046ad2e9b1faf8a7b4d3aa8011fc6a77
---
 drivers/usb/gadget/function/f_mass_storage.c | 11 +++++++++++
 drivers/usb/gadget/function/storage_common.h |  3 +++
 2 files changed, 14 insertions(+)

diff --git a/drivers/usb/gadget/function/f_mass_storage.c b/drivers/usb/gadget/function/f_mass_storage.c
index c265a1f62fc1..022cbdcd6915 100644
--- a/drivers/usb/gadget/function/f_mass_storage.c
+++ b/drivers/usb/gadget/function/f_mass_storage.c
@@ -2728,7 +2728,12 @@ static void _fsg_common_free_buffers(struct fsg_buffhd *buffhds, unsigned n)
 	if (buffhds) {
 		struct fsg_buffhd *bh = buffhds;
 		while (n--) {
+#if defined(CONFIG_SOC_SPACEMIT_K1X)
+			put_page(bh->page);
+			bh->page = NULL;
+#else
 			kfree(bh->buf);
+#endif
 			++bh;
 		}
 		kfree(buffhds);
@@ -2752,7 +2757,13 @@ int fsg_common_set_num_buffers(struct fsg_common *common, unsigned int n)
 		bh->next = bh + 1;
 		++bh;
 buffhds_first_it:
+#if defined(CONFIG_SOC_SPACEMIT_K1X)
+		bh->page = alloc_pages(GFP_KERNEL | __GFP_COMP | __GFP_NOWARN | GFP_DMA32,
+			get_order(FSG_BUFLEN));
+		bh->buf = page_address(bh->page);
+#else
 		bh->buf = kmalloc(FSG_BUFLEN, GFP_KERNEL);
+#endif
 		if (unlikely(!bh->buf))
 			goto error_release;
 	} while (--i);
diff --git a/drivers/usb/gadget/function/storage_common.h b/drivers/usb/gadget/function/storage_common.h
index 0a544a82cbf8..81cb567ca30e 100644
--- a/drivers/usb/gadget/function/storage_common.h
+++ b/drivers/usb/gadget/function/storage_common.h
@@ -141,6 +141,9 @@ enum fsg_buffer_state {
 };
 
 struct fsg_buffhd {
+#if defined(CONFIG_SOC_SPACEMIT_K1X)
+	struct page			*page;
+#endif
 	void				*buf;
 	enum fsg_buffer_state		state;
 	struct fsg_buffhd		*next;
-- 
2.47.0

